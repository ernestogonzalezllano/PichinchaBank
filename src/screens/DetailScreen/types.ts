import {NativeStackScreenProps} from '@react-navigation/native-stack';
import {RootStackParamList} from '../../types/screensTypes';

export type DetailScreenProps = NativeStackScreenProps<
  RootStackParamList,
  'Detail'
>;
