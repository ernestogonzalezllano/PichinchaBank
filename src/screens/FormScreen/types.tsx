import {NativeStackScreenProps} from '@react-navigation/native-stack';
import {RootStackParamList} from '../../types/screensTypes';
import {GapsNames} from '../../types/api';

export type FormScreenProps = NativeStackScreenProps<
  RootStackParamList,
  'Form'
>;

export type FormGaps = {
  [key in GapsNames]: string | undefined;
};
