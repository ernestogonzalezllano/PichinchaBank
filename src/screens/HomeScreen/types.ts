import {NativeStackScreenProps} from '@react-navigation/native-stack';
import {RootStackParamList} from '../../types/screensTypes';

export type HomeProps = NativeStackScreenProps<RootStackParamList, 'Home'>;
