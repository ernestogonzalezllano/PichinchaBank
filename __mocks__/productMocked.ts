export const productMocked = {
  id: '12345',
  date_release: '2024-11-11',
  date_revision: '2025-11-11',
  description: 'test react testing library',
  logo: 'https://www.visa.com.ec/dam/VCOM/regional/lac/SPA/Default/Pay%20With%20Visa/Tarjetas/visa-signature-400x225.jpg',
  name: 'test test',
};
