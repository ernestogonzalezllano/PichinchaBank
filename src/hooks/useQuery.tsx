import {useCallback, useEffect, useState} from 'react';
import {BASE_URL} from '../constants';
import {ApiRoutes} from '../types/api';
import {useFocusEffect} from '@react-navigation/native';

interface Props {
  url: ApiRoutes;
  refetchOnMount?: boolean;
  options?: {
    body?: {[key: string]: any};
    method?: 'POST' | 'PUT' | 'DELETE' | 'GET';
    params?: {[key: string]: any};
  };
}

export const useQuery = <T,>({url, refetchOnMount = true, options}: Props) => {
  const [data, setData] = useState<T>();
  const [error, setError] = useState<string>();
  const [loading, setLoading] = useState<boolean>(refetchOnMount ?? false);

  const refetch = useCallback(() => {
    setData(undefined);
    setError(undefined);
    setLoading(true);
    const body = options?.body ? JSON.stringify(options.body) : undefined;
    const params = options?.params
      ? '?' + new URLSearchParams(options.params).toString()
      : '';

    fetch(`${BASE_URL}${url}${params}`, {
      ...options,
      body,
      headers: {
        authorId: '10',
        'Content-Type': 'application/json',
      },
    })
      .then(response => {
        if (!response.ok) {
          throw new Error('Something was wrong');
        }
        return response.text().then(textResponse => {
          let result = textResponse;
          try {
            result = JSON.parse(textResponse);
          } catch {}
          return result as T;
        });
      })
      .then(result => {
        setData(result);
      })
      .catch(() => {
        setError('Error al obtener datos de la API');
      })
      .finally(() => {
        setLoading(false);
      });
  }, [options, url]);

  const myFunction = useCallback(() => {
    if (url && refetchOnMount) {
      refetch();
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    return () => {
      setData(undefined);
      setError(undefined);
    };
  }, []);

  useFocusEffect(myFunction);
  return {data, error, refetch, loading};
};
