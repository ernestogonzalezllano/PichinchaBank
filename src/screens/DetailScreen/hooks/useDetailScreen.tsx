import {useState} from 'react';
import {useQuery} from '../../../hooks/useQuery';
import {Product} from '../../../types/api';

export const useDetailScreen = ({product}: {product: Readonly<Product>}) => {
  const [openModal, setOpenModal] = useState(false);
  const {refetch, loading, error, data} = useQuery<string>({
    url: '/bp/products',
    options: {method: 'DELETE', params: {id: product.id}},
    refetchOnMount: false,
  });
  return {openModal, setOpenModal, refetch, loading, error, data};
};
