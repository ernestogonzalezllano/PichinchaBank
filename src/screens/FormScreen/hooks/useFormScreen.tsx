import {useState} from 'react';
import {FormGaps} from '../types';
import {useQuery} from '../../../hooks/useQuery';
import {validations} from '../utils';
import {Product} from '../../../types/api';
import {useValidateId} from './useValidateId';

const defaultFormData = {
  id: undefined,
  date_release: undefined,
  date_revision: undefined,
  description: undefined,
  logo: 'https://www.visa.com.ec/dam/VCOM/regional/lac/SPA/Default/Pay%20With%20Visa/Tarjetas/visa-signature-400x225.jpg',
  name: undefined,
};

export const useFormScreen = ({
  product,
}: {
  product: Readonly<Product | undefined>;
}) => {
  const isEdit = !!product;
  const [formData, setFormData] = useState<FormGaps>(
    product ?? defaultFormData,
  );
  const [errors, setErrors] = useState({});

  const {data, refetch, error, loading} = useQuery<Product>({
    url: '/bp/products',
    refetchOnMount: false,
    options: {body: formData, method: product ? 'PUT' : 'POST'},
  });

  const {validateId} = useValidateId({id: formData.id, isEdit});

  const handleSubmit = () => {
    const validate = validations(formData, validateId);
    if (Object.keys(validate).length) {
      setErrors(validate);
    } else {
      refetch();
    }
  };
  return {
    setFormData,
    setErrors,
    errors,
    data,
    error,
    loading,
    handleSubmit,
    formData,
    defaultFormData,
    isEdit,
  };
};
