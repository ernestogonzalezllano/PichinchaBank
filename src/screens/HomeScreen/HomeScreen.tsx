import React, {Suspense} from 'react';
import {FlatList, StyleSheet, View} from 'react-native';
import {CardList} from '../../components/CardList';
import {ButtonTouchable} from '../../components/ButtonTouchable';

import {Input} from '../../components/Input';
import {EmptyProducts} from '../../components/EmptyProducts';
import {HomeProps} from './types';
import {SkeletonLoading} from './components/SkeletonLoading';
import {useHomeScreen} from './hooks/useHomeScreen';

export const HomeScreen = ({navigation}: HomeProps) => {
  const {
    setSearch,
    deferredQuery,
    data,
    loading,
    handleGoToDetail,
    handleGoToForm,
  } = useHomeScreen({navigation});
  return (
    <View style={styles.container}>
      <Input
        placeholder="Buscar..."
        onChange={text => {
          setSearch(text);
        }}
      />
      <Suspense>
        {loading ? (
          <SkeletonLoading />
        ) : (
          <FlatList
            data={
              deferredQuery
                ? data?.filter(({id}) => id.includes(deferredQuery!!))
                : data
            }
            keyExtractor={item => item.id}
            ListEmptyComponent={EmptyProducts}
            renderItem={item => (
              <CardList {...item} handleOnPress={handleGoToDetail} />
            )}
            contentContainerStyle={styles.contentContainerStyle}
          />
        )}
      </Suspense>
      <ButtonTouchable label="Agregar" onPress={handleGoToForm} />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    gap: 5,
  },
  input: {
    height: 40,
    padding: 10,
    borderWidth: 1,
    borderRadius: 5,
    marginVertical: 20,
  },
  contentContainerStyle: {
    flexGrow: 1,
    gap: 5,
  },
});
