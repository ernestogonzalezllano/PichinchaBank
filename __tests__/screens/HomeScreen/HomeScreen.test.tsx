import * as React from 'react';
import * as useQuery from '../../../src/hooks/useQuery';
import {
  fireEvent,
  render,
  screen,
  waitFor,
} from '@testing-library/react-native';
import {NavigationContainer} from '@react-navigation/native';

import {HomeScreen} from '../../../src/screens/HomeScreen/HomeScreen';
import {HomeProps} from '../../../src/screens/HomeScreen/types';

import {productMocked as originalProductMocked} from '../../../__mocks__/productMocked';

const productMocked = {...originalProductMocked};
const navigate = jest.fn();

const props = {
  navigation: {
    navigate,
  },
  route: {},
} as unknown as HomeProps;

describe('HomeScreen', () => {
  test('Should match with snapshot', async () => {
    const component = renderScreen().component.toJSON();
    expect(component).toMatchSnapshot();
  });
  test('Should show skeleton when is loading', async () => {
    renderScreen(undefined, undefined, true);
    expect(screen.getAllByTestId('HomeCardSkeleton')).toHaveLength(10);
  });
  test('Should filter when search', async () => {
    renderScreen([productMocked, {...productMocked, id: '5678'}]);
    fireEvent.changeText(screen.getByPlaceholderText('Buscar...'), '5678');
    await waitFor(() => {
      expect(screen.getAllByText(productMocked.name)).toHaveLength(1);
    });
  });
  test('Should go to ScreenForm when "Agregar" button is pressed', async () => {
    renderScreen();
    fireEvent.press(screen.getByText('Agregar'));
    expect(navigate).toHaveBeenCalledWith('Form');
  });
  test('Should go to ScreenDetail when CardProduct is pressed', async () => {
    renderScreen();
    fireEvent.press(screen.getByText(productMocked.name));
    expect(navigate).toHaveBeenCalledWith('Detail', productMocked);
  });
});

const renderScreen = (
  data = [productMocked],
  error = undefined,
  loading = false,
) => {
  const refetch = jest.fn();

  jest.spyOn(useQuery, 'useQuery').mockReturnValue({
    data,
    error,
    refetch,
    loading,
  });
  const component = render(
    <NavigationContainer>
      <HomeScreen {...props} />
    </NavigationContainer>,
  );
  return {component, refetch};
};
