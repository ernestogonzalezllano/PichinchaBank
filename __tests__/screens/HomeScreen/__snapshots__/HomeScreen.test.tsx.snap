// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen Should match with snapshot 1`] = `
<View
  style={
    {
      "flex": 1,
      "gap": 5,
    }
  }
>
  <View>
    <TextInput
      editable={true}
      onChangeText={[Function]}
      placeholder="Buscar..."
      style={
        {
          "borderColor": "#eaecf3",
          "borderRadius": 5,
          "borderWidth": 1,
          "height": 40,
          "padding": 10,
        }
      }
    />
  </View>
  <RCTScrollView
    ListEmptyComponent={[Function]}
    contentContainerStyle={
      {
        "flexGrow": 1,
        "gap": 5,
      }
    }
    data={
      [
        {
          "date_release": "2024-11-11",
          "date_revision": "2025-11-11",
          "description": "test react testing library",
          "id": "12345",
          "logo": "https://www.visa.com.ec/dam/VCOM/regional/lac/SPA/Default/Pay%20With%20Visa/Tarjetas/visa-signature-400x225.jpg",
          "name": "test test",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "borderColor": "#eaecf3",
              "borderRadius": 5,
              "borderWidth": 1,
              "flexDirection": "row",
              "opacity": 1,
              "padding": 10,
            }
          }
        >
          <View
            style={
              {
                "flex": 1,
              }
            }
          >
            <Text
              style={
                {
                  "fontWeight": "bold",
                }
              }
            >
              test test
            </Text>
            <Text>
              ID: 
              12345
            </Text>
          </View>
          <View
            style={
              {
                "justifyContent": "center",
                "width": 20,
              }
            }
          >
            <Image
              source={
                {
                  "testUri": "../../../assets/right.png",
                }
              }
              style={
                {
                  "height": 30,
                  "position": "relative",
                  "resizeMode": "contain",
                  "width": "auto",
                }
              }
            />
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#ffdd04",
        "borderRadius": 5,
        "opacity": 1,
        "padding": 10,
      }
    }
  >
    <Text
      style={
        {
          "backgroundColor": "#ffdd04",
          "fontWeight": "bold",
        }
      }
    >
      Agregar
    </Text>
  </View>
</View>
`;
